File: day8/haskell/part1.hs
import Data.List (sortBy)
import qualified Data.Map as Map
import Data.Ord (comparing)

type Point = (Int, Int, Int)

-- Split string by delimiter
splitOn :: Char -> String -> [String]
splitOn _ [] = [""]
splitOn delim (c:cs)
  | c == delim = "" : rest
  | otherwise  = (c : head rest) : tail rest
  where rest = splitOn delim cs

-- Parse a line like "162,817,812" into (162, 817, 812)
parsePoint :: String -> Point
parsePoint s = (read x, read y, read z)
  where [x, y, z] = splitOn ',' s

-- Euclidean distance between two 3D points
distance :: Point -> Point -> Double
distance (x1, y1, z1) (x2, y2, z2) = sqrt $ dx*dx + dy*dy + dz*dz
  where
    dx = fromIntegral (x1 - x2)
    dy = fromIntegral (y1 - y2)
    dz = fromIntegral (z1 - z2)

-- Union-Find using a Map
-- parent map: index -> parent index
type Parent = Map.Map Int Int

-- Find root with path compression (returns new map and root)
find :: Parent -> Int -> (Parent, Int)
find parent x =
  case Map.lookup x parent of
    Nothing -> (Map.insert x x parent, x)
    Just p | p == x -> (parent, x)
           | otherwise ->
               let (parent', root) = find parent p
               in (Map.insert x root parent', root)

-- Union two elements
union :: Parent -> Int -> Int -> Parent
union parent x y =
  let (parent1, rootX) = find parent x
      (parent2, rootY) = find parent1 y
  in if rootX == rootY
     then parent2
     else Map.insert rootX rootY parent2

-- Apply n unions from sorted pairs
applyUnions :: Parent -> [(Int, Int, Double)] -> Int -> Parent
applyUnions parent _ 0 = parent
applyUnions parent [] _ = parent
applyUnions parent ((i, j, _):rest) n = applyUnions (union parent i j) rest (n - 1)

-- Get all circuit sizes
getCircuitSizes :: Parent -> Int -> [Int]
getCircuitSizes parent n = Map.elems counts
  where
    -- Find root for each node
    (finalParent, roots) = foldr (\i (p, rs) ->
        let (p', r) = find p i
        in (p', r:rs)) (parent, []) [0..n-1]
    -- Count nodes per root
    counts = foldr (\r m -> Map.insertWith (+) r 1 m) Map.empty roots

main :: IO ()
main = do
  content <- readFile "input.txt"
  let points = map parsePoint $ lines content
      n = length points

      -- Generate all pairs with distances
      pairs = [(i, j, distance (points !! i) (points !! j))
              | i <- [0..n-1], j <- [i+1..n-1]]

      -- Sort by distance
      sortedPairs = sortBy (comparing (\(_,_,d) -> d)) pairs

      -- Apply 1000 unions
      parent = applyUnions Map.empty sortedPairs 1000

      -- Get sizes and sort descending
      sizes = reverse $ sortBy compare $ getCircuitSizes parent n

      -- Multiply top 3
      result = (sizes !! 0) * (sizes !! 1) * (sizes !! 2)

  print result

File: day8/go/part1.go
package main

import (
	"bufio"
	"fmt"
	"math"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Point struct {
	x, y, z int
}

type Pair struct {
	i, j int
	dist float64
}

// Union-Find data structure
type UnionFind struct {
	parent []int
	rank   []int
}

func NewUnionFind(n int) *UnionFind {
	parent := make([]int, n)
	rank := make([]int, n)
	for i := 0; i < n; i++ {
		parent[i] = i
	}
	return &UnionFind{parent, rank}
}

func (uf *UnionFind) Find(x int) int {
	if uf.parent[x] != x {
		uf.parent[x] = uf.Find(uf.parent[x]) // path compression
	}
	return uf.parent[x]
}

func (uf *UnionFind) Union(x, y int) {
	px, py := uf.Find(x), uf.Find(y)
	if px == py {
		return
	}
	// union by rank
	if uf.rank[px] < uf.rank[py] {
		uf.parent[px] = py
	} else if uf.rank[px] > uf.rank[py] {
		uf.parent[py] = px
	} else {
		uf.parent[py] = px
		uf.rank[px]++
	}
}

func distance(a, b Point) float64 {
	dx := float64(a.x - b.x)
	dy := float64(a.y - b.y)
	dz := float64(a.z - b.z)
	return math.Sqrt(dx*dx + dy*dy + dz*dz)
}

func main() {
	file, _ := os.Open("input.txt")
	defer file.Close()

	var points []Point
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := scanner.Text()
		parts := strings.Split(line, ",")
		x, _ := strconv.Atoi(parts[0])
		y, _ := strconv.Atoi(parts[1])
		z, _ := strconv.Atoi(parts[2])
		points = append(points, Point{x, y, z})
	}

	n := len(points)

	// Calculate all pairwise distances
	var pairs []Pair
	for i := 0; i < n; i++ {
		for j := i + 1; j < n; j++ {
			pairs = append(pairs, Pair{i, j, distance(points[i], points[j])})
		}
	}

	// Sort by distance
	sort.Slice(pairs, func(a, b int) bool {
		return pairs[a].dist < pairs[b].dist
	})

	// Connect 1000 closest pairs using Union-Find
	uf := NewUnionFind(n)
	connections := 1000
	for i := 0; i < connections && i < len(pairs); i++ {
		uf.Union(pairs[i].i, pairs[i].j)
	}

	// Count circuit sizes
	sizes := make(map[int]int)
	for i := 0; i < n; i++ {
		root := uf.Find(i)
		sizes[root]++
	}

	// Get all sizes and sort descending
	var sizeList []int
	for _, size := range sizes {
		sizeList = append(sizeList, size)
	}
	sort.Sort(sort.Reverse(sort.IntSlice(sizeList)))

	// Multiply top 3
	result := sizeList[0] * sizeList[1] * sizeList[2]
	fmt.Println(result)
}
